lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py

```diff
@@ -0,0 +1,189 @@
+    password: vcenter_password
+    switch_name: dvSwitch
+    version: version3
+    state: present
+'''
+
+try:
+    from pyVmomi import vim, vmodl
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware import (HAS_PYVMOMI,
+                                         connect_to_api,
+                                         find_datacenter_by_name,
+                                         find_dvs_by_name,
+                                         vmware_argument_spec,
+                                         wait_for_task
+                                         )
+
+
+class VMwareDVSwitchNIOC(object):
+
+    def __init__(self, module):
+        self.module = module
+        self.dvs = None
+        self.switch_name = self.module.params['switch_name']
+        self.version = self.module.params['version']
+        self.state = self.module.params['state']
+        self.content = connect_to_api(module)
+
+    def process_state(self):
+        dvs_states = {
+            'absent': {
+                'present': self.state_disable_nioc,
+                'absent': self.state_exit_unchanged,
+            },
+            'present': {
+                'update': self.state_update_nioc,
+                'present': self.state_exit_unchanged,
+                'absent': self.state_enable_nioc,
+            }
+        }
+        dvs_states[self.state][self.check_nioc_state()]()
+
+    def state_exit_unchanged(self):
+        self.module.exit_json(changed=False)
+
+    def state_disable_nioc(self):
+        changed = True
+        result = "Disable NIOC"
+
+        if not self.module.check_mode:
+            task = self.dvs.EnableNetworkResourceManagement(enable=False)
+
+        self.module.exit_json(changed=changed, result=str(result))
+
+    def state_enable_nioc(self):
+        changed = True
+        result = "Enable NIOC"
+
+        if not self.module.check_mode:
+            task = self.dvs.EnableNetworkResourceManagement(enable=True)
+            self.set_nioc_version()
+
+        self.module.exit_json(changed=changed, result=str(result))
+
+    def state_update_nioc(self):
+        changed = False
+        result = ""
+
+        if not self.module.check_mode:
+            self.set_nioc_version()
+
+        self.module.exit_json(changed=changed, result=str(result))
+
+    def set_nioc_version(self):
+        valid_versions = ['version2', 'version3']
+        if self.version not in valid_versions:
+            raise ValueError('version "{0}" is invalid. Accepted values: {1}'.format(self.version, valid_versions))
+
+        upgrade_spec = vim.DistributedVirtualSwitch.ConfigSpec()
+        upgrade_spec.configVersion = self.dvs.config.configVersion
+        upgrade_spec.networkResourceControlVersion = self.version
+        task = self.dvs.ReconfigureDvs_Task(spec=upgrade_spec)
+
+        changed, result = wait_for_task(task)
```

Comment:
might want to wrap `wait_for_task` with try except block to handle exception.

Reply:
And exit with a proper failure message if it fails.